contrib/cmake/build3rdParty/build3rdParty.cpp

```diff
@@ -0,0 +1,273 @@
+#include <iostream>
+#include <filesystem>
+#include <vector>
+#include <map>
+#include <string>
+#include <fstream>
+
+using namespace std;
+
+
+#if (__cplusplus >= 201703L && (!defined(__GNUC__) || (__GNUC__*100+__GNUC_MINOR__) >= 800))
```

Comment:
If we're checking for `__cplusplus >= 2017` that means we should have access to `__has_include`, might be worth simplifying some of the logic that way? 

```cpp
#if (__cplusplus >= 201703L)
#if __has_include(<filesystem)>
#include <filesystem>
namespace fs = std::filesystem
#else // __has_include(<filesystem)>
#include <experimental/filesystem>
namespace fs = std::experimental::filesystem
#else // (__cplusplus >= 201703L)
// ...
#endif 
```

Reply:
This particular block of #ifs etc is our current method of figuring out which compilers have some version of std::filesystem or std::experimental::filesystem that we can use - it's been through many, many revisions and it's quite difficult to get right across all the platforms, I'd be wary of editing it unless there's a case it doesn't cover.